core/utils/utils.go

```diff
@@ -418,46 +418,48 @@ func WaitGroupChan(wg *sync.WaitGroup) <-chan struct{} {
 // channel is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (ctx context.Context, cancel context.CancelFunc) {
-	ctx, cancel = context.WithCancel(parentCtx)
-
-	go func() {
-		select {
-		case <-chStop:
-		case <-ctx.Done():
-		}
-		cancel()
-	}()
-
-	return ctx, cancel
+func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithCancel(parentCtx))
 }
 
 // ContextFromChan creates a context that finishes when the provided channel
 // receives or is closed.
-// When channel closes, the ctx.Err() will always be context.Canceled
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
 func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		select {
-		case <-chStop:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	return ctx, cancel
+	return StopChan(chStop).NewCtx()
 }
 
-// ContextFromChanWithDeadline creates a context with a deadline that finishes when the provided channel
+// ContextFromChanWithTimeout creates a context with a timeout that finishes when the provided channel
 // receives or is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func ContextFromChanWithDeadline(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func ContextFromChanWithTimeout(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithTimeout(context.Background(), timeout))
+}
+
+type StopChan <-chan struct{}
```

Comment:
does this need to be external?

StopChan is vague and the fact that it's backed by a channel seems like an implementation detail. How about ContextCanceller or ContextController or

Reply:
It makes it more flexible with existing `stopCh chan strtuct{}`s, which we have a lot of.